astropy/utils/data.py

```diff
@@ -1327,43 +1400,303 @@ def _get_download_cache_locs():
 
 # the cache directory must be locked before any writes are performed.  Same for
 # the hash shelve, so this should be used for both.
-def _acquire_download_cache_lock():
-    """
-    Uses the lock directory method.  This is good because `mkdir` is
-    atomic at the system call level, so it's thread-safe.
-    """
-
+#
+# In fact the shelve, if you're using gdbm (the default on Linux), it can't
+# be open for reading if someone has it open for writing. So we need to lock
+# access to the shelve even for reading.
+@contextlib.contextmanager
+def _cache_lock():
```

Comment:
FIXME: make the cache usable in a read-only mode - if people have their cache on a read-only filesystem as in #8677 they should be able to read from but not write to the cache; in this case it is not necessary to acquire the lock to prevent contention for the shelve object (multiple readers are permitted).

Reply:
Done. With less pain than I expected.